cpp/fetcher/fetcher.cc

```diff
@@ -0,0 +1,254 @@
+        current->next_ = move(current->next_->next_);
+      }
+    }
+
+    switch (current->state_) {
+      case Range::HAVE:
+        VLOG(1) << "at offset " << index << ", we have " << current->size_
+                << " entries";
+        break;
+
+      case Range::FETCHING:
+        VLOG(1) << "at offset " << index << ", fetching " << current->size_
+                << " entries";
+        ++num_fetch;
+        break;
+
+      case Range::WANT:
+        VLOG(1) << "at offset " << index << ", we want " << current->size_
+                << " entries";
+
+        // If the range is bigger than the maximum batch size, split it.
+        if (current->size_ > kMaxBatchSize) {
+          current->next_.reset(new Range(Range::WANT,
+                                         current->size_ - kMaxBatchSize,
+                                         move(current->next_)));
+          current->size_ = kMaxBatchSize;
+        }
+
+        StartSingleFetch(current, index);
+        ++num_fetch;
+
+        break;
+    }
+
+    if (num_fetch >= kMaxFetches) {
+      break;
+    }
+  }
+
+  VLOG(1) << "num_fetch: " << num_fetch << ", " << entries_.get();
+}
+
+
+void FetchState::StartSingleFetch(Range* current, int64_t index) {
+  const int64_t end_index(index + current->size_ - 1);
+  VLOG(1) << "fetching from offset " << index << " to " << end_index;
+
+  // In-flight fetches should prevent our task from completing.
+  task_->AddHold();
+
+  vector<AsyncLogClient::Entry>* const retval(
+      new vector<AsyncLogClient::Entry>);
+  peer_->client().GetEntries(index, end_index, retval,
+                             bind(&FetchState::SingleFetchDone, this, index,
+                                  current, retval, _1));
+
+  current->state_ = Range::FETCHING;
+}
+
+
+void FetchState::SingleFetchDone(int64_t index, Range* range,
+                                 vector<AsyncLogClient::Entry>* retval,
+                                 AsyncLogClient::Status status) {
+  task_->executor()->Add(bind(&FetchState::BlockingSingleFetchDone, this,
+                              index, range, retval, status));
+}
+
+
+void FetchState::BlockingSingleFetchDone(int64_t index, Range* range,
+                                         vector<AsyncLogClient::Entry>* retval,
+                                         AsyncLogClient::Status status) {
+  CHECK_GT(retval->size(), 0);
+
+  unique_ptr<vector<AsyncLogClient::Entry>> retval_deleter(retval);
+  TaskHold hold(task_);
+  task_->RemoveHold();
+
+  {
+    lock_guard<mutex> lock(range->lock_);
+
+    if (status != AsyncLogClient::OK) {
+      LOG(INFO) << "error fetching entries at index " << index << ": "
+                << status;
+      range->state_ = Range::WANT;
+      return;
+    }
+
+    VLOG(1) << "received " << retval->size() << " entries at offset " << index;
+    int64_t processed(0);
+    for (const auto& entry : *retval) {
+      LoggedCertificate cert;
+      if (!cert.CopyFromClientLogEntry(entry)) {
```

Comment:
I think this probably merits at least a LOG(WARNING) or something, in all likelihood we're going to get wedged here on all subsequent runs (modulo a duff peer returning garbage etc.)